### ISSUE TITLE
Encoding issues with GHC's error messages

### ISSUE THREAD

**Issue by regnat** [tweag/rules_haskell#523]
We've recently started encountering encoding issues with the error messages of GHC (after a conjoint update of `rules_haskell`, `ghc` and `bazel`, so not sure who's the culprit). GHC apparently now tries to print non-ascii unicode characters, but these aren't correctly printed:

![image](https://user-images.githubusercontent.com/7226587/50220460-a7647d00-0392-11e9-98f8-a6f6a8e2bdd9.png)

This doesn't appear in NixOS machines nor in the nixos/nix docker image where GHC fallbacks to ASCII-only messages (but seems to appear everywhere else).

/cc-ing @guibou and @aherrmann who had the same issue

**Comment by judah** [comment]
Those look like the color codes:
https://downloads.haskell.org/~ghc/master/users-guide/using.html#ghc-flag--fdiagnostics-color=%E2%9F%A8always|auto|never%E2%9F%A9

Whether GHC does that is controlled by `-fdiagnostics-color`.  It defaults to `auto`, in which case GHC looks at the value of `$TERM` to decide whether they're supported.  As a workaround, you can probably set `-fdiagnostics-color=never` in the `haskell_toolchain`'s `compiler_flags` to disable them for now.

What operating system and terminal emulator(s) are you using?

**Comment by guibou** [comment]
As a workaround I used `locale = "POSIX"` in `haskell_toolchain` to "fix" the issue. Build should not rely on locale, so it is fine most of the time.

But a proper solution may be interesting.

cc @ghorn which is interested too.

**Comment by guibou** [comment]
This appears for me now on my nixos machine.

What are you thoughts on forcing `local = "POSIX"` by default and opening a new (low priority) issue about restoring nice "unicode" aware GHC error message.

I rather prefer having correct ASCII error message than incorrect ones.

Note: the tests of rules_haskell depends on utf-8 locale for some `readFile`. I'm convinced that it is a bug, `readFile` should not be used when you know the encoding of the file you are reading, but only when you suppose that the file you are reading is written in the same locale as the one used on the system.

**Comment by mboes** [comment]
> What are you thoughts on forcing local = "POSIX" by default and opening a new (low priority) issue about restoring nice "unicode" aware GHC error message?

The problem with this approach is that it's dangerous, due to GHC limitations. GHC really wants a unicode locale, or you're asking for trouble. See commercialhaskell/stack#793 and commercialhaskell/stack#4294. Furthermore, doing so now means that no one can get localized error messages from their toolchains.

**Comment by feuerbach** [comment]
> Note: the tests of rules_haskell depends on utf-8 locale for some `readFile`. I'm convinced that it is a bug, `readFile` should not be used when you know the encoding of the file you are reading, but only when you suppose that the file you are reading is written in the same locale as the one used on the system.

Where does 'the other' locale even come from? If I am running a utf-8 locale, why is not everything in utf-8?

**Comment by guibou** [comment]
@feuerbach 

> Where does 'the other' locale even come from? If I am running a utf-8 locale, why is not everything in utf-8?

You may have an utf8 system, but you may also have softwares which stores data in another encoding, because it is and old software, or it comes from a place where utf16 is the norm, because it is a raw dump of something which internally is not utf8.

Depending on the system local is a not too wrong guess when you have no other information. But if the software is controling how files are written to the disk, it must also controls how it'll read them, and in this case you don't want the system local to mess with that.

**Comment by feuerbach** [comment]
@guibou point taken. I was talking from the perspective of the user who experiences the original bug reported here.

Are you saying that `readFile` is causing the bug, or that you consider its use a separate bug?

If it's the former, then I am confused as to how relying on the locale can possibly cause this. If anything, it seems to be ignoring the locale and using an 8-bit encoding or something.

If it's the latter, then could you open a separate issue?

**Comment by jcpetruzza** [comment]
We were experiencing the same thing, but it didn't seem to be a haskell issue since a dummy genrule that printed unicode text and failed would exhibit the same behaviour. FWIW, we were on bazel 0.20 from nixpkgs; after bumping nixpkgs and getting bazel 0.21, the problem went away.

**Comment by Profpatsch** [comment]
@regnat Do we still see this issue with bazel 0.21?

**Comment by guibou** [comment]
Fixed here. I'm closing. Feel free to reopen if that's still an issue for you on recent bazel.
